src/pool/srv_target.c

```diff
@@ -1344,3 +1344,53 @@ ds_pool_tgt_prop_update(struct ds_pool *pool, struct pool_iv_prop *iv_prop)
 	pool->sp_reclaim = iv_prop->pip_reclaim;
 	return 0;
 }
+
+/**
+ * Query the cached pool map. If the cached version is <= in->tmi_map_version,
+ * reply -DER_ALREADY.
+ */
+void
+ds_pool_tgt_query_map_handler(crt_rpc_t *rpc)
+{
+	struct pool_tgt_query_map_in   *in = crt_req_get(rpc);
+	struct pool_tgt_query_map_out  *out = crt_reply_get(rpc);
+	struct ds_pool_hdl	       *hdl;
+	struct ds_pool		       *pool;
+	struct pool_buf		       *buf;
+	unsigned int			version;
+	int				rc;
+
+	D_DEBUG(DB_TRACE, DF_UUID": handling rpc %p\n",
+		DP_UUID(in->tmi_op.pi_uuid), rpc);
+
+	hdl = ds_pool_hdl_lookup(in->tmi_op.pi_hdl);
+	if (hdl == NULL) {
+		rc = -DER_NO_HDL;
+		goto out;
+	}
+
+	/* Inefficient; better invent some zero-copy IV APIs. */
+	pool = hdl->sph_pool;
+	ABT_rwlock_rdlock(pool->sp_lock);
+	version = pool_map_get_version(pool->sp_map);
+	if (version > in->tmi_map_version)
+		rc = pool_buf_extract(pool->sp_map, &buf);
+	else
+		rc = -DER_ALREADY;
+	ABT_rwlock_unlock(pool->sp_lock);
```

Comment:
this is stream 0, and no update here, do we need lock here?

Reply:
Technically we do not, but let's do so in all cases. There have been multiple errors recently where people probably take xstream 0 code as a (wrong) example and missing locks on other xstreams.